drivers/adc/aducm3029/aducm3029_adc.c

```diff
@@ -0,0 +1,178 @@
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <drivers/adc/adi_adc.h>
+#include "aducm3029_adc.h"
+#include "error.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "util.h"
+
+#define ADUCM3029_DEV_ID 0
+
+struct adc_desc {
+	uint8_t		dev_mem[ADI_ADC_MEMORY_SIZE];
+	ADI_ADC_HANDLE	dev;
+	uint32_t	ch_mask;
+};
+
+int32_t aducm3029_adc_update_active_channels(struct adc_desc *desc,
+		uint32_t mask)
+{
+	if (!desc)
+		return -EINVAL;
+
+	desc->ch_mask = mask;
+
+	return SUCCESS;
+}
+
+int32_t aducm3029_adc_read(struct adc_desc *desc, uint16_t *buff,
+			   uint32_t nb_samples)
+{
+	ADI_ADC_BUFFER	adi_buff, *pbuff;
+	int32_t		ret;
+	uint32_t	i;
+
+	if (!desc)
+		return -EINVAL;
+
+	for (i = 0; i < nb_samples; i++) {
+		adi_buff.nChannels = desc->ch_mask;
+		adi_buff.pDataBuffer = buff + i * hweight8(desc->ch_mask);
+		adi_buff.nNumConversionPasses = 1;
+		adi_buff.nBuffSize = hweight8(desc->ch_mask) * sizeof(uint16_t);
+		ret = adi_adc_SubmitBuffer(desc->dev, &adi_buff);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+		ret = adi_adc_Enable(desc->dev, true);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+		ret = adi_adc_GetBuffer(desc->dev, &pbuff);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+		ret = adi_adc_Enable(desc->dev, false);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+	}
+
+	return SUCCESS;
+}
+
+static int inline count_bits(uint32_t ch_mask)
```

Comment:
Maybe use hweight8 instead. It seems it accepts 32bit variables. Maybe the name of it is not quite meaningful.

Reply:
I forgot to remove this function

Reply:
I was not used